src/lsm/k_way_merge.zig

```diff
@@ -105,10 +105,13 @@ pub fn KWayMergeIterator(
             if (stream_peek(it.context, root)) |key| {
                 it.keys[0] = key;
                 it.down_heap();
-            } else {
-                it.swap(0, it.k - 1);
-                it.k -= 1;
-                it.down_heap();
+            } else |err| switch (err) {
+                error.Pending => return null,
+                error.Empty => {
+                    it.swap(0, it.k - 1);
+                    it.k -= 1;
+                    it.down_heap();
+                },
             }
```

Comment:
So Pending should do the same as the successful `if |key|` branch to ensure it's peek'd again?

Reply:
It can't because it doesn't have a `key` to use.

I think we need to move the whole heap update immediately before we remove a value. e.g.

```
if (stream_peek(it.context, root)) |key| {
                it.keys[0] = key;
                it.down_heap();
            } else {
                it.swap(0, it.k - 1);
                it.k -= 1;
                it.down_heap();
            } else |err| switch (err) {
                error.Pending => return null,
                error.Empty => {
                    it.swap(0, it.k - 1);
                    it.k -= 1;
                    it.down_heap();
                },
            }
```

must all move before

```
            const root = it.streams[0];
            // We know that each input iterator is sorted, so we don't need to compare the next
            // key on that iterator with the current min/max.
            const value = stream_pop(it.context, root);
```